File: c/heap.c
#include <alloca.h>
#include <error.h>

#include "common.h"

unsigned int maxu(unsigned int a, unsigned int b) { return a > b ? a : b; }

typedef struct HeapNode {
	double value;
	struct HeapNode* left;
	struct HeapNode* right;
	int empty;
} HeapNode;

HeapNode* node_new_sides(double value, HeapNode* left, HeapNode* right) {
	HeapNode* ret = malloc(sizeof(HeapNode));
	ret->value = value;
	ret->left = left;
	ret->right = right;
	ret->empty = 0;

	return ret;
}

HeapNode* node_new(double value) { return node_new_sides(value, 0, 0); }

unsigned int node_depth(HeapNode* node) {
	unsigned int ldepth = node->left ? node_depth(node->left) + 1 : 1;
	unsigned int rdepth = node->right ? node_depth(node->right) + 1 : 1;

	return maxu(ldepth, rdepth);
}

void node_push(HeapNode* node, double new_value) {
	if (new_value >= node->value) {
		if (!node->left) {
			node->left = node_new(new_value);
		} else if (!node->right) {
			node->right = node_new(new_value);
		} else {
			HeapNode* expand_side =
			    node_depth(node->left) < node_depth(node->right) ? node->left
			                                                     : node->right;
			node_push(expand_side, new_value);
		}
	} else {
		double old_value = node->value;
		HeapNode* old_right = node->right;
		HeapNode* old_left = node->left;

		node->value = new_value;
		node->right = 0;
		node->left = node_new_sides(old_value, old_left, old_right);
	}
}

double node_pop(HeapNode* node) {
	double ret = node->value;

	if (node->left && node->right) {
		HeapNode* pop_side =
		    node->left->value < node->right->value ? node->left : node->right;
		node->value = node_pop(pop_side);
	} else if (node->left) {
		node->value = node->left->value;
		node->right = node->left->right;
		node->left = node->left->left;
	} else if (node->right) {
		node->value = node->right->value;
		node->left = node->right->left;
		node->right = node->right->right;
	} else {
		node->empty = 1;
	}

	if (node->left && node->left->empty) {
		free(node->left);
		node->left = 0;
	}
	if (node->right && node->right->empty) {
		free(node->right);
		node->right = 0;
	}

	return ret;
}

size_t node_count(HeapNode* node) {
	if (node) {
		return 1 + node_count(node->left) + node_count(node->right);
	} else {
		return 0;
	}
}

const size_t OUT_SIZE = 1000;
void node_to_string(HeapNode* node, char* out, size_t out_len) {
	if (node) {
		char* l_out = alloca(sizeof(char) * OUT_SIZE);
		char* r_out = alloca(sizeof(char) * OUT_SIZE);

		node_to_string(node->left, l_out, OUT_SIZE);
		node_to_string(node->right, r_out, OUT_SIZE);

		snprintf(out, out_len, "HeapNode(%f, %s, %s)", node->value, l_out,
		         r_out);
	} else {
		snprintf(out, out_len, "—");
	}
}

void node_print(HeapNode* node) {
	char* out = alloca(sizeof(char) * OUT_SIZE);

	node_to_string(node, out, OUT_SIZE);

	puts(out);
	puts("\n");
}

typedef HeapNode* Heap;

void heap_push(Heap* h, double val) {
	if (*h) {
		node_push(*h, val);
	} else {
		*h = node_new(val);
	}
}

double heap_pop(Heap* h) {
	if (*h) {
		double ret = node_pop(*h);
		if ((*h)->empty) {
			free(*h);
			*h = 0;
		}
		return ret;
	} else {
		error_at_line(-1, -1, __FILE__, __LINE__,
		              "Attempted to pop empty heap");
		return 0;
	}
}

inline int heap_empty(Heap* h) { return *h == 0; }

inline Heap heap_new() { return 0; }

inline size_t heap_count(Heap* h) { return node_count(*h); }

void heap_free(Heap* h) {
	while (!heap_empty(h)) {
		heap_pop(h);
	}
}

inline void heap_print(Heap* h) { node_print(*h); }

double act(int size) {
	double* l = calloc(size, sizeof(double));

	for (int i = 0; i < size; i++) {
		l[i] = get_number();
	}
#ifdef DEBUG
	print_list(l, size);
#endif

	double start = CLOCK();
	Heap h = heap_new();
	for (int i = 0; i < size; i++) {
		heap_push(&h, l[i]);
	}
#ifdef DEBUG
	heap_print(&h);
#endif

	free(l);
	double* out = calloc(heap_count(&h), sizeof(double));

	for (int i = 0; i < size; i++) {
		out[i] = heap_pop(&h);
	}

	double end = CLOCK();

	fprintf(stderr, "%f\n", l[rand() % size]);

#ifdef DEBUG
	print_list(out, size);
#endif
	free(out);

	return end - start;
}

int main(int argc, char** argv) { return main_g(argc, argv, act); }

File: python/heap.py
#!/usr/bin/env python3
import pprint
import random as rand
from dataclasses import dataclass
from typing import Optional

from common import get_number
from common import run_bench
from common import timef

pp = pprint.PrettyPrinter().pprint


@dataclass
class HeapNode:
    value: any
    left: Optional["HeapNode"] = None
    right: Optional["HeapNode"] = None
    __empty: bool = False

    def push(self, new_value: any):
        if new_value >= self.value:
            if self.left is None:
                self.left = HeapNode(new_value)
            elif self.right is None:
                self.right = HeapNode(new_value)
            else:
                expand_side = (
                    self.left if self.left.depth < self.right.depth else self.right
                )
                expand_side.push(new_value)
        else:
            old_value = self.value
            old_right = self.right
            old_left = self.left

            self.value = new_value
            self.right = None
            self.left = HeapNode(old_value, old_left, old_right)

    def pop(self) -> any:
        ret = self.value
        if self.left and self.right:
            pop_side = self.left if self.left.value < self.right.value else self.right
            self.value = pop_side.pop()
        elif self.left:
            self.value = self.left.value
            self.right = self.left.right
            self.left = self.left.left
        elif self.right:
            self.value = self.right.value
            self.left = self.right.left
            self.right = self.right.right
        else:
            self.__empty = True

        if self.left and self.left.empty:
            self.left = None
        if self.right and self.right.empty:
            self.right = None

        return ret

    @property
    def depth(self) -> int:
        ldepth = self.left.depth + 1 if self.left else 1
        rdepth = self.right.depth + 1 if self.right else 1

        return max(ldepth, rdepth)

    @property
    def empty(self) -> bool:
        return self.__empty


@dataclass
class Heap:
    head_node: Optional[HeapNode] = None

    def push(self, val: any):
        if self.head_node:
            self.head_node.push(val)
        else:
            self.head_node = HeapNode(val)

    def pop(self):
        if self.head_node:
            ret = self.head_node.pop()
            if self.head_node.empty:
                self.head_node = None
            return ret
        else:
            raise ValueError("Popped empty heap")

    def empty():
        return self.head_node is not None

    def pop_all(self):
        while self.head_node:
            yield self.pop()


def act(size: int) -> int:
    unsorted = [get_number() for i in range(size)]

    def sort():
        h = Heap()
        for e in unsorted:
            h.push(e)

        return [e for e in h.pop_all()]

    time, s = timef(lambda: sort())

    if len(s) < 50:
        pp(s)
    else:
        print(s[rand.randint(0, len(s) - 1)])

    return time


if __name__ == "__main__":
    run_bench(
        lambda size: act(size),
        description="Sorting floats using a heap",
        size_help="Number of floats to sort",
    )
